test/e2e_node/container_lifecycle_test.go

```diff
@@ -398,6 +398,184 @@ var _ = SIGDescribe(framework.WithNodeConformance(), "Containers Lifecycle", fun
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "container-image-update-onfailure",
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyOnFailure,
+				Containers: []v1.Container{
+					{
+						Name:            regular1,
+						Image:           agnhostImage,
+						ImagePullPolicy: v1.PullAlways,
+						Command: ExecCommand(regular1, execCommand{
+							Delay:    30,
+							ExitCode: 0,
+						}),
+					},
+				},
+			},
+		}
+
+		preparePod(podSpec)
+
+		ginkgo.It("should not restart when the image is updated", func(ctx context.Context) {
+			client := e2epod.NewPodClient(f)
+			podSpec = client.Create(ctx, podSpec)
+
+			ginkgo.By("running the pod", func() {
+				err := e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, podSpec.Name, podSpec.Namespace)
+				framework.ExpectNoError(err)
+			})
+
+			ginkgo.By("updating the image", func() {
+				client.Update(ctx, podSpec.Name, func(pod *v1.Pod) {
+					pod.Spec.Containers[0].Image = "fakeimage"
+				})
+
+			})
+
+			ginkgo.By("verifying that the container does not restart", func() {
+				err := e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, podSpec.Name, podSpec.Namespace)
+				framework.ExpectNoError(err)
+				podSpec, err = client.Get(ctx, podSpec.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				err = e2epod.WaitForPodCondition(ctx, f.ClientSet, podSpec.Namespace, podSpec.Name, "the container terminated regularly", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+					containerStatus := pod.Status.ContainerStatuses[0]
+					return containerStatus.State.Terminated != nil && containerStatus.State.Terminated.ExitCode == 0, nil
+				})
+				framework.ExpectNoError(err)
+
+				results := parseOutput(ctx, f, podSpec)
+				framework.ExpectNoError(results.HasNotRestarted(regular1))
+			})
+		})
+	})
+
+	ginkgo.When("A pod is running a regular container with restartPolicy=Always", func() {
+		regular1 := "regular-1"
+		bufferSeconds := int64(20)
+		exitSeconds := 60
+
+		podSpec := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "container-image-update-always",
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyAlways,
+				Containers: []v1.Container{
+					{
+						Name:            regular1,
+						Image:           agnhostImage,
+						ImagePullPolicy: v1.PullAlways,
+						Command: ExecCommand(regular1, execCommand{
+							Delay:    exitSeconds,
+							ExitCode: 0,
+						}),
+					},
+				},
+			},
+		}
+
+		preparePod(podSpec)
+
+		ginkgo.It("should restart when the image is updated", func(ctx context.Context) {
+
+			client := e2epod.NewPodClient(f)
+			podSpec = client.Create(ctx, podSpec)
+
+			ginkgo.By("running the pod", func() {
+				err := e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, podSpec.Name, podSpec.Namespace)
+				framework.ExpectNoError(err)
+			})
+
+			ginkgo.By("updating the image", func() {
+				client.Update(ctx, podSpec.Name, func(pod *v1.Pod) {
+					pod.Spec.Containers[0].Image = "fakeimage"
```

Comment:
We have a curated list of images which are okay to use in tests: https://github.com/kubernetes/kubernetes/blob/master/test/images/.permitted-images

We don't have different tags for the same image. So for this test, I think switching to `registry.k8s.io/pause` is fine. There's also `registry.k8s.io/e2e-test-images/busybox`, but it might get (be?) superseded by `agnhost` and I wouldn't add more usages of it.

Reply:
Instead of `fakeimage`, use `invalid.registry.k8s.io/invalid/alpine`?

Reply:
Pause is probably suitable for this, the main difference in recent versions is windows support.

However we want to keep it minimal. So maybe the one pause image and the one agnhost image, instead of multiple tags for those images. Which would also mean we are at the latest windows support for both.

Sorry I lost this thread in all the other github notifications.